src/Parser.js

```diff
@@ -308,10 +309,10 @@ class Parser {
      *
      * @return {?ParseNode}
      */
-    parseAtom() {
+    parseAtom(breakOnTokenText) {
         // The body of an atom is an implicit group, so that things like
         // \left(x\right)^2 work correctly.
-        const base = this.parseImplicitGroup();
+        const base = this.parseImplicitGroup(breakOnTokenText);
```

Comment:
It might be worth adding a comment somewhere as to why it's okay not to pass this parameter down to other methods below. The reason seems to be subtle: as `handleSupSubscript` can call `parseGroup` which can call `parseExpression`, but `parseExpression` is only called in `parseGroup` only if a `{` or `[` is encountered, opening a deeper layer. But I haven't yet done an exhaustive check.

I'm also fine with completely deferring this if we intend to formalize the various non-obvious invariants in this file some point to aid comprehensibility so that they aren't violated in the future.

Reply:
I'll add a comment.  `handleSupSubscript` should always be parsing non-optional groups afaict because `x^[2]` isn't valid (or rather it's valid it's treated as `x^{[}2]`).